File: csharp/solution/Webshop.Service/Products/ProductService.cs
namespace Webshop.Service.Products;

public interface IProductService
{
    Product? GetProduct(string id);
    int GetProductInventory(string productId);
    void SubtractFromProductInventory(string productId, int quantity);
}

public class ProductService : IProductService
{
    private readonly IProductRepository _productRepository;

    public ProductService(IProductRepository productRepository)
    {
        _productRepository = productRepository;
    }

    public Product? GetProduct(string id)
    {
        return _productRepository.GetProduct(id);
    }

    public int GetProductInventory(string productId)
    {
        return _productRepository.GetProductInventory(productId);
    }

    public void SubtractFromProductInventory(string productId, int quantity)
    {
        var existingInventory = _productRepository.GetProductInventory(productId);
        var newInventory = existingInventory - quantity;
        _productRepository.UpdateProductInventory(productId, newInventory);
    }
}

File: java/exercises/src/main/java/Webshop/Service/Products/ProductService.java
package Webshop.Service.Products;

interface IProductService {
    Product getProduct(String id);
    int getProductInventory(String productId);
    void subtractFromProductInventory(String productId, int quantity);
}

public class ProductService implements IProductService {
    private final IProductRepository productRepository;

    public ProductService(IProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    @Override
    public Product getProduct(String id) {
        return productRepository.getProduct(id);
    }

    @Override
    public int getProductInventory(String productId) {
        return productRepository.getProductInventory(productId);
    }

    @Override
    public void subtractFromProductInventory(String productId, int quantity) {
        int existingInventory = productRepository.getProductInventory(productId);
        int newInventory = existingInventory - quantity;
        productRepository.updateProductInventory(productId, newInventory);
    }
}

